### PR TITLE
[Instrumentation.StackExchangeRedis] Support when IConnectionMultiplexer is added with keyed service

### LINKED ISSUE #1451: Support Redis instrumentation when IConnectionMultiplexer is added with keyed service
# Issue with OpenTelemetry.Instrumentation.StackExchangeRedis

**Is this a feature request or a bug?**

* [x] Feature Request
* [ ] Bug

Currently there's two way to have Redis instrumentation. With passing the `IConnectionMultiplexer` instance to the method or By registering `IConnectionMultiplexer` to the IoC Container.

In .Net 8 there's support to keyed services in IoC Container. It would be nice to have support for instrumentation for `IConnectionMultiplexer` that is registered as a keyed service.

I suggest adding this API

```csharp
public class TracerProviderBuilderExtensions
{
   public static TracerProviderBuilder AddRedisInstrumentation(
        this TracerProviderBuilder builder,
        string? name,
        IConnectionMultiplexer? connection,
        Action<StackExchangeRedisInstrumentationOptions>? configure,
        object? serviceKey)
}
```
And this line need to change so that if `serviceKey` is available get the `IConnectionMultiplexer` with `servicekey`.

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/c3a4f335f5a1f4fc784d44cc99b5bf44e872a66c/src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs#L135

Also to support this the OpenTelemetry.Instrumentation.StackExchangeRedis project need to add .NET 8 to its target framekwork.

I'm willing to send a PR for this feature if approved.

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes #1451.

## Changes

Added a new overload to AddRedisInstrumentation which gets serviceKey in input and get IConnectionMultiplexer with a keyed service.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #
* [x] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/.publicApi/net8.0/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+OpenTelemetry.Instrumentation.StackExchangeRedis.StackExchangeRedisInstrumentation
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/.publicApi/net8.0/PublicAPI.Unshipped.txt:1]: I think that the contract is common for all targets. This file (and the second one under .ne8 folder) should be removed.

Ref: #1711

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 2]
```diff
@@ -119,8 +156,15 @@ public static class TracerProviderBuilderExtensions
...
-
+#if NET8_0_OR_GREATER
+                connection ??= serviceKey == null
+                    ? sp.GetService<IConnectionMultiplexer>()
+                    : sp.GetKeyedService<IConnectionMultiplexer>(serviceKey);
+#else
+                connection ??= serviceKey == null
+                    ? sp.GetService<IConnectionMultiplexer>()
+                    : throw new InvalidOperationException("Redis instrumentation with service key is supported in .NET 8.0 and above.");
+#endif
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:167]: Based on https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1451#issuecomment-1834172834. There is no need to conditional compilation. It is indirect reference by `"OpenTelemetry.Api.ProviderBuilderExtensions v1.8.1` 

```suggestion
                connection ??= serviceKey == null
                    ? sp.GetService<IConnectionMultiplexer>()
                    : sp.GetKeyedService<IConnectionMultiplexer>(serviceKey);
```

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
-    <TargetFrameworks>net6.0;netstandard2.0;net462</TargetFrameworks>
+    <TargetFrameworks>net8.0;net6.0;netstandard2.0;net462</TargetFrameworks>
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:4]: Please revert if this is true https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1451#issuecomment-1834172834

File: test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj [hunk 4]
```diff
@@ -18,6 +18,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="OpenTelemetry.Api.ProviderBuilderExtensions" Version="$(OpenTelemetryCoreLatestVersion)" />
```
Comment (Kielek) [inline test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj:21]: I do not think that this reference is needed. It is transitive dependency by 

```xml
<ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.StackExchangeRedis\OpenTelemetry.Instrumentation.StackExchangeRedis.csproj"
```

**Comment by Kielek** [comment]
@eerhardt, please check this PR, you were reviewing the first version.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj [hunk 5]
```diff
@@ -22,6 +22,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="8.0.0" />
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:25]: This line is not needed. See https://www.nuget.org/packages/OpenTelemetry.Api.ProviderBuilderExtensions/1.8.1#dependencies-body-tab

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 6]
```diff
@@ -40,7 +40,42 @@ public static class TracerProviderBuilderExtensions
...
+
+    /// <summary>
+    /// Enables automatic data collection of outgoing requests to Redis.
+    /// </summary>
+    /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
+    /// <param name="name">Optional name which is used when retrieving options.</param>
+    /// <param name="serviceKey">Optional service key used to retrieve the <see cref="IConnectionMultiplexer"/> to instrument from the <see cref="IServiceProvider" />.</param>
+    /// <param name="configure">Callback to configure options.</param>
+    /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+    public static TracerProviderBuilder AddRedisInstrumentation(
```
Comment (eerhardt) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:69]: Do these new overloads have tests covering them?

**Review by eerhardt — APPROVED** [review]
Looks good to me. Could probably use a few new tests.

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 7]
```diff
@@ -54,12 +89,12 @@ public static class TracerProviderBuilderExtensions
     /// <param name="configure">Callback to configure options.</param>
     /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
     public static TracerProviderBuilder AddRedisInstrumentation(
-        this TracerProviderBuilder builder,
-        Action<StackExchangeRedisInstrumentationOptions> configure)
+                this TracerProviderBuilder builder,
+                Action<StackExchangeRedisInstrumentationOptions> configure)
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:93]:
```suggestion
        this TracerProviderBuilder builder,
        Action<StackExchangeRedisInstrumentationOptions> configure)
```

Just reverting whitespace change.

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Pull request merged**